File: src/lib/scrapers/recruitee.ts
/**
 * Recruitee ATS Scraper
 *
 * Fetches job listings from Recruitee public API.
 * API: https://{company}.recruitee.com/api/offers
 */

import {
  stripHtml,
  extractSalary,
  retryWithBackoff,
  generateSourceId,
} from './utils';
import type { ScrapedJobData } from './greenhouse';

// ==================== Types ====================

export interface RecruiteeJob {
  id: number;
  title: string;
  city?: string;
  location?: string;
  remote?: boolean;
  hybrid?: boolean;
  on_site?: boolean;
  careers_url: string;
  careers_apply_url: string;
  created_at: string;
  published_at?: string;
  requirements?: string;
  employment_type_code?: string;
  experience_code?: string;
  category_code?: string;
  department?: string;
  min_hours?: number;
  max_hours?: number;
  salary?: {
    min?: number;
    max?: number;
    period?: string;
    currency?: string;
  };
  locations?: Array<{
    id: number;
    name: string;
    city?: string;
    state?: string;
    country?: string;
    country_code?: string;
  }>;
  translations?: {
    en?: {
      title?: string;
      description?: string;
      requirements?: string;
    };
  };
}

export interface RecruiteeResponse {
  offers: RecruiteeJob[];
}

// ==================== Constants ====================

const RATE_LIMIT_DELAY_MS = 1000; // 1s between requests

// ==================== Scraper Functions ====================

/**
 * Fetch all jobs from a Recruitee company
 */
export async function fetchRecruiteeJobs(
  slug: string,
  companyName: string
): Promise<ScrapedJobData[]> {
  const url = `https://${slug}.recruitee.com/api/offers`;

  const response = await retryWithBackoff(
    async () => {
      const res = await fetch(url, {
        headers: {
          'Accept': 'application/json',
          'User-Agent': 'GenzJobs/1.0 (job aggregator)',
        },
      });

      if (!res.ok) {
        if (res.status === 404) {
          throw new Error(`Recruitee company not found: ${slug}`);
        }
        if (res.status === 429) {
          throw new Error('Rate limited by Recruitee API');
        }
        throw new Error(`Recruitee API error: ${res.status} ${res.statusText}`);
      }

      return res.json() as Promise<RecruiteeResponse>;
    },
    {
      maxRetries: 3,
      initialDelayMs: 1000,
      maxDelayMs: 30000,
    }
  );

  const jobs: ScrapedJobData[] = [];

  for (const job of response.offers) {
    const scraped = parseRecruiteeJob(job, slug, companyName);
    if (scraped) {
      jobs.push(scraped);
    }
  }

  return jobs;
}

/**
 * Parse a single Recruitee job into our standard format
 */
function parseRecruiteeJob(
  job: RecruiteeJob,
  slug: string,
  companyName: string
): ScrapedJobData | null {
  // Get description from translations or requirements
  const rawDescription = job.translations?.en?.description || job.requirements || '';
  const description = stripHtml(rawDescription);

  // Skip jobs without meaningful description
  if (!description || description.length < 50) {
    return null;
  }

  // Build location string
  const location = buildLocationString(job);

  // Detect remote
  const remote = job.remote === true ||
    location?.toLowerCase().includes('remote') ||
    false;

  // Extract salary - use structured data if available, otherwise parse from description
  let salaryMin: number | null = null;
  let salaryMax: number | null = null;
  let salaryCurrency: string | null = null;
  let salaryPeriod: string | null = null;

  if (job.salary?.min || job.salary?.max) {
    salaryMin = job.salary.min || null;
    salaryMax = job.salary.max || null;
    salaryCurrency = job.salary.currency || null;
    salaryPeriod = mapSalaryPeriod(job.salary.period);
  } else {
    const extracted = extractSalary(description);
    if (extracted) {
      salaryMin = extracted.min;
      salaryMax = extracted.max;
      salaryCurrency = extracted.currency;
      salaryPeriod = extracted.period;
    }
  }

  // Parse job type
  const jobType = parseEmploymentType(job.employment_type_code);

  // Parse posted date
  const postedAt = job.published_at
    ? new Date(job.published_at)
    : new Date(job.created_at);

  return {
    sourceId: generateSourceId('recruitee', slug, String(job.id)),
    source: 'recruitee',
    title: job.title,
    company: companyName,
    location,
    remote,
    description,
    applyUrl: job.careers_apply_url || job.careers_url,
    salaryMin,
    salaryMax,
    salaryCurrency,
    salaryPeriod,
    jobType,
    department: job.department || job.category_code || null,
    postedAt,
  };
}

/**
 * Build location string from Recruitee job data
 */
function buildLocationString(job: RecruiteeJob): string | null {
  // Try locations array first
  if (job.locations && job.locations.length > 0) {
    const loc = job.locations[0];
    const parts = [loc.city, loc.state, loc.country].filter(Boolean);
    if (parts.length > 0) {
      return parts.join(', ');
    }
  }

  // Fall back to location field
  if (job.location) {
    return job.location;
  }

  // Fall back to city
  if (job.city) {
    return job.city;
  }

  return null;
}

/**
 * Map Recruitee salary period to our format
 */
function mapSalaryPeriod(period?: string): string | null {
  if (!period) return null;

  const lower = period.toLowerCase();
  if (lower === 'yearly' || lower === 'annual' || lower === 'year') {
    return 'yearly';
  }
  if (lower === 'monthly' || lower === 'month') {
    return 'monthly';
  }
  if (lower === 'hourly' || lower === 'hour') {
    return 'hourly';
  }
  if (lower === 'weekly' || lower === 'week') {
    return 'weekly';
  }

  return null;
}

/**
 * Parse Recruitee employment type into standardized job type
 */
function parseEmploymentType(typeCode?: string): string | null {
  if (!typeCode) return null;

  const lower = typeCode.toLowerCase();

  if (lower.includes('fulltime') || lower.includes('full_time') || lower === 'permanent') {
    return 'full-time';
  }
  if (lower.includes('parttime') || lower.includes('part_time')) {
    return 'part-time';
  }
  if (lower.includes('contract') || lower.includes('contractor') || lower.includes('temporary')) {
    return 'contract';
  }
  if (lower.includes('intern')) {
    return 'internship';
  }
  if (lower.includes('freelance')) {
    return 'contract';
  }

  return null;
}

/**
 * Map Recruitee experience level to our classification
 */
export function mapExperienceLevel(experienceCode?: string): string | null {
  if (!experienceCode) return null;

  const lower = experienceCode.toLowerCase();

  if (lower === 'entry_level' || lower === 'internship' || lower === 'student') {
    return 'ENTRY';
  }
  if (lower === 'mid_level' || lower === 'associate') {
    return 'MID';
  }
  if (lower === 'senior_level' || lower === 'senior') {
    return 'SENIOR';
  }
  if (lower === 'director' || lower === 'executive' || lower === 'manager') {
    return 'EXECUTIVE';
  }

  return null;
}

/**
 * Validate a Recruitee company exists
 */
export async function validateRecruiteeCompany(slug: string): Promise<boolean> {
  try {
    const url = `https://${slug}.recruitee.com/api/offers`;
    const response = await fetch(url, {
      headers: {
        'User-Agent': 'GenzJobs/1.0 (job aggregator)',
      },
    });
    return response.ok;
  } catch {
    return false;
  }
}

/**
 * Get rate limit delay for Recruitee
 */
export function getRecruiteeRateLimitDelay(): number {
  return RATE_LIMIT_DELAY_MS;
}

File: scraper/scrapers/recruitee.py
import httpx
import logging
from typing import List, Optional
from datetime import datetime

from .ats_base import ATSBaseScraper
from models import ScrapedJob
from companies import RECRUITEE_COMPANIES

logger = logging.getLogger(__name__)


class RecruiteeScraper(ATSBaseScraper):
    """Scraper for Recruitee ATS boards.
    API: GET https://{slug}.recruitee.com/api/offers
    Returns {offers: [...]}. Has structured salary and experience level fields.
    """

    def __init__(self):
        super().__init__("recruitee", RECRUITEE_COMPANIES)

    async def fetch_company_jobs(self, company_name: str, slug: str) -> List[ScrapedJob]:
        jobs: List[ScrapedJob] = []
        url = f"https://{slug}.recruitee.com/api/offers"

        async with httpx.AsyncClient() as client:
            response = await client.get(url, timeout=30.0)
            response.raise_for_status()
            data = response.json()

        for raw_job in data.get("offers", []):
            job = self._parse_job(raw_job, company_name, slug)
            if job:
                jobs.append(job)

        return jobs

    def _parse_job(self, raw_job: dict, company_name: str, slug: str) -> Optional[ScrapedJob]:
        job_id = raw_job.get("id")
        title = raw_job.get("title", "").strip()
        if not job_id or not title:
            return None

        # Status check - only published offers
        status = raw_job.get("status", "")
        if status and status != "published":
            return None

        # Description: try translations.en first, then description
        description = ""
        translations = raw_job.get("translations", {})
        if translations and isinstance(translations, dict):
            en = translations.get("en", {})
            if isinstance(en, dict):
                description = en.get("description", "")
        if not description:
            description = raw_job.get("description", "")
        description = self.strip_html(description) if "<" in description else description

        # Location
        location = raw_job.get("location", "")
        city = raw_job.get("city", "")
        country = raw_job.get("country", "")
        if not location:
            location_parts = [p for p in [city, country] if p]
            location = ", ".join(location_parts)

        # Remote
        remote = raw_job.get("remote", False) or self.detect_remote(description, location)

        # Job type from employment_type_code
        job_type = self.normalize_job_type(raw_job.get("employment_type_code"))

        # Salary
        salary_min = raw_job.get("min_salary")
        salary_max = raw_job.get("max_salary")
        salary_currency = raw_job.get("salary_currency", "USD")

        # Convert to int if present
        if salary_min is not None:
            try:
                salary_min = int(salary_min)
            except (ValueError, TypeError):
                salary_min = None
        if salary_max is not None:
            try:
                salary_max = int(salary_max)
            except (ValueError, TypeError):
                salary_max = None

        # Apply URL
        careers_url = raw_job.get("careers_url") or raw_job.get("url")
        apply_url = careers_url or f"https://{slug}.recruitee.com/o/{raw_job.get('slug', job_id)}"

        # Posted date
        posted_at = None
        published = raw_job.get("published_at") or raw_job.get("created_at")
        if published:
            try:
                posted_at = datetime.fromisoformat(str(published).replace("Z", "+00:00"))
            except (ValueError, TypeError):
                pass

        return self.make_job(
            external_id=str(job_id),
            title=title,
            company_name=company_name,
            description=description,
            apply_url=apply_url,
            location=location,
            remote=remote,
            job_type=job_type,
            salary_min=salary_min,
            salary_max=salary_max,
            salary_currency=salary_currency,
            posted_at=posted_at,
        )
